core/state/snapshot/generate.go

```diff
@@ -159,169 +178,313 @@ func journalProgress(db ethdb.KeyValueWriter, marker []byte, stats *generatorSta
 	rawdb.WriteSnapshotGenerator(db, blob)
 }
 
-// generate is a background thread that iterates over the state and storage tries,
-// constructing the state snapshot. All the arguments are purely for statistics
-// gathering and logging, since the method surfs the blocks as they arrive, often
-// being restarted.
-func (dl *diskLayer) generate(stats *generatorStats) {
-	// If a database wipe is in operation, wait until it's done
-	if stats.wiping != nil {
-		stats.Log("Wiper running, state snapshotting paused", common.Hash{}, dl.genMarker)
-		select {
-		// If wiper is done, resume normal mode of operation
-		case <-stats.wiping:
-			stats.wiping = nil
-			stats.start = time.Now()
-
-		// If generator was aborted during wipe, return
-		case abort := <-dl.genAbort:
-			abort <- stats
-			return
+// proveRange proves the state segment with particular prefix is "valid".
+// The iteration start point will be assigned if the iterator is restored from
+// the last interruption. Max will be assigned in order to limit the maximum
+// amount of data involved in each iteration.
+func (dl *diskLayer) proveRange(root common.Hash, tr *trie.SecureTrie, prefix []byte, kind string, origin []byte, max int, onValue func([]byte) ([]byte, error)) ([][]byte, [][]byte, []byte, bool, error) {
+	var (
+		keys  [][]byte
+		vals  [][]byte
+		count int
+		last  []byte
+		proof = rawdb.NewMemoryDatabase()
+		iter  = dl.diskdb.NewIterator(prefix, origin)
+	)
+	for iter.Next() && count < max {
+		key := iter.Key()
+		if len(key) != len(prefix)+common.HashLength {
+			continue
+		}
+		if !bytes.HasPrefix(key, prefix) {
+			continue
+		}
+		last = common.CopyBytes(key[len(prefix):])
+		keys = append(keys, common.CopyBytes(key[len(prefix):]))
+
+		if onValue == nil {
+			vals = append(vals, common.CopyBytes(iter.Value()))
+		} else {
+			val, err := onValue(common.CopyBytes(iter.Value()))
+			if err != nil {
+				log.Debug("Failed to convert the flat state", "kind", kind, "key", common.BytesToHash(key[len(prefix):]), "error", err)
+				return nil, nil, last, false, err
+			}
+			vals = append(vals, val)
+		}
+		count += 1
+	}
+	// Generate the Merkle proofs for the first and last element
+	if origin == nil {
+		origin = common.Hash{}.Bytes()
+	}
+	if err := tr.Prove(origin, 0, proof); err != nil {
+		log.Debug("Failed to prove range", "kind", kind, "origin", origin, "err", err)
+		return nil, nil, last, false, err
+	}
+	if last != nil {
+		if err := tr.Prove(last, 0, proof); err != nil {
+			log.Debug("Failed to prove range", "kind", kind, "last", last, "err", err)
+			return nil, nil, last, false, err
 		}
 	}
-	// Create an account and state iterator pointing to the current generator marker
-	accTrie, err := trie.NewSecure(dl.root, dl.triedb)
+	// Verify the state segment with range prover, ensure that all flat states
+	// in this range correspond to merkle trie.
+	_, _, _, cont, err := trie.VerifyRangeProof(root, origin, last, keys, vals, proof)
```

Comment:
Since you ignore the db here, it would be neat to use a version of the range prover which uses the stacktrie instead of the full trie. 
OTOH it doesn't matter a whole lot, since you're only using a few hundred key/values per iteration

Reply:
Actually the stacktrie is not suitable for range prove. Stacktrie expects all the inputs from the most left side and to the most right side. It won't accept the segment in the middle, even the segment is continuous.